### PR TITLE
Ignore empty aggregates elements when binding aggregate direction

### PR DESCRIPTION
**Pull request by jackkoenig**
Previously, including an empty aggregate in a Bundle would cause
a MixedDirectionAggregateException because it has no elements and thus
doesn't have a direction

**Related issue**: 

**Type of change**: bug report and fix

**Impact**: API addition? Stuff that doesn't work (but should) now does

**Development Phase**: implementation

**Release Notes**
Empty aggregates no longer cause MixedDirectionAggregateException when included as members of Bundles

h/t @davidbiancolin for identifying and reducing the problem

### CODE REVIEW

**Comment by jackkoenig** [comment]
I grant that the given empty Bundle test case looks a little bit funky, but I'm not really sure how you distinguish it from more reasonable looking yet functionally equivalent cases:
```scala
val io = IO(new Bundle {
  val in = Input(UInt(8.W))
  val a = new Bundle { }
  val b = new Bundle {
    val foo = if (false) Some(Input(UInt(8.W))) else None
  }
  val c = MyCustomRecord(List.fill(0)(Output(UInt(8.W))))
})
```
While I see the argument for making `val a` above illegal, I don't think it's possible to do so without making `b` and `c` illegal which would be very problematic. We need a test case for empty Records which is what I was using `new Bundle { }` to accomplish.

I agree that Vec directionality logic should run on the sample element if only to save unnecessary computation. I can change that as part of this PR or it could be a follow on?

**Comment by ducky64** [comment]
I think it also makes sense to include the sample_element change on this PR to minimize the number of edge cases. It should also be a pretty small change?

**Comment by jackkoenig** [comment]
Yeah I'll include the sample_element change in this PR. I don't entirely understand what edge cases it mitigates though? It looks like purely a performance improvement to me.

**Comment by ducky64** [comment]
It should treat directioned
`val x = Vec(0, Input(UInt(8.W)))`
differently from undirectioned
`val x = Vec(0, UInt(8.W))`

I think (but might be worth adding a test case for?) existing logic already differentiates directioned
`val x = Input(Vec(0, UInt(8.W)))`
from undirectioned
`val x = Vec(0, UInt(8.W))`

**Comment by jackkoenig** [comment]
Got it, I'll add that

**Comment by ducky64** [comment]
I think architecturally the cleaning solution is just to bind sample_element. It occupies this weird space between hardware (or representing hardware at least) and not hardware (it's kind of a ghost object). I think having a clear SampleElementBinding or something would be the cleanest if its implementation isn't too onerous. Such a solution would also be ideal for some of the other edge cases mentioned (eg optional Bundle elements), though I think the Scala execution style precludes that (since None optional Bundle elements are never actually evaluated).

**Comment by jackkoenig** [comment]
Alright this should be good to go. I added a special binding for `sample_element` and tweaking the `topBindingOpt` to recognize it so that `sample_element`s don't get treated as hardware.

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 1]
```diff
@@ -258,9 +264,10 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
     _binding = Some(target)
   }
 
-  private[core] def topBindingOpt: Option[TopBinding] = _binding.map {
-    case ChildBinding(parent) => parent.topBinding
-    case bindingVal: TopBinding => bindingVal
+  private[core] def topBindingOpt: Option[TopBinding] = _binding.flatMap {
+    case ChildBinding(parent) => parent.topBindingOpt // Parent could be a SampleElement itself
+    case bindingVal: TopBinding => Some(bindingVal)
+    case _: SampleElementBinding[_] => None
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Consider making this assert out, since this really shouldn't happen ever, right? (sample element shouldn't be exposed to the user)

File: chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala [hunk 2]
```diff
@@ -25,14 +25,22 @@ sealed abstract class Aggregate extends Data {
...
+      case vec: Vec[_] =>
+        // Use direction of sample_element, must bind it to get an ActualDirection
+        vec.sample_element.bind(SampleElementBinding(vec), resolvedDirection)
+        Set(vec.sample_element.direction)
+      case rec: Record =>
+        getElements.collect({
+          case elt: Element => elt.direction
+          case vec: Vec[_] => vec.direction
+          // Don't check the directions of empty Records (see https://github.com/freechipsproject/chisel3/pull/946)
+          case rec: Record if rec.getElements.nonEmpty => rec.direction
```
Comment (ducky64) [inline chiselFrontend/src/main/scala/chisel3/core/Aggregate.scala]: I can't quite say what the larger implications of this will be, but the inspecting the contents of this's elements seems like an odd code smell. Perhaps it might make sense to have an ActualDirection,NoDirection, because that's kind of what we're getting at here fundamentally?

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 1 above)
Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: `topBindingOpt` is called by [`Module` for allocating ids](https://github.com/freechipsproject/chisel3/blob/5c32cd87b85e2607e558f286cb1f8fd9604e4c83/chiselFrontend/src/main/scala/chisel3/core/UserModule.scala#L76). This change is explicitly making anything that has a `SampleElementBinding` parent (since you can have sample elements that are themselves aggregates) return None for it's top binding. The whole point of this change is to make sure that sample element isn't exposed to the user.

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: It looks like `topBindingOpt` is used in a lot of places to indicate that something is an actual piece of hardware. I'm happy to wrap it in some way but in any case the `SampleElementBinding` will come up so it either needs to map to `None` in `topBindingOpt` or in some hypothetical new method that wraps `topBindingOpt` but only returns bindings that represent real hardware

Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Oh I see.

Longer term, it might make sense to not have every HasId that gets instantiate try to add itself to the parent, since in some cases it doesn't make sense like for types (unbound), literals, and sample elements.

Comment (jackkoenig) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Yeah I totally agree with the longer term vision. I'll add some comments to explain what `topBindingOpt` is and how it's used so that it's clear why `SampleElementBinding` returns `None`

**Comment by chick** [comment]
@jackkoenig this popped up for me to review but it's a little hard to tell the state. It looks to me like it's awaiting some more changes before requiring review.  Is that right?

**Comment by ducky64** [comment]
@jackkoenig made a few changes:
- Wrote a stricter definition in comments for ConstrainedBinding, that the FIRRTL generated is referenceable (needed for the StrongEnum self-binding test - though cleaner solutions are appreciated since this contributes to the can't-have-SampleElementBinding-assert-out-in-topBinding)
- Add a TODO about SampleElementBinding in topBindingOpt, which can be refactored another day
- Add ActualDirection.Empty, which is ignored when inspecting Aggregate child directions
- Restructure Aggregate.bind into ActualDirection.fromChildren, Aggregate.bind (much simplified), Record.bind (which catches the MixedDirectionAggregateException and discards it in compatibility mode; put the compatibility hack in the most specific place), Vec.bind
- This restructuring gets rid of a few scalastyle:ignore annotations - yay!
- Vec.bind looks at the sample element exclusively to determine container direction

One interesting observation that came up: the behavior of Vec apply is inconsistent between the dynamic (UInt arg) and static (Int arg) case: in the dynamic case, it requires the parent Vec to be bound, but in the static case, there are no checks (it's just a Vector access). Are there cases where an unbound Vec needs to be indexed? (this occurs at least once in chiselTests, but only to determine SpecifiedDirection of a Vec element - and in this case it might make more sense to allow access to the sample element (or an unbound clone of it) which would also better support inspecting into zero-length Vecs.

We might also consider cloning from sample_element instead of calling gen, since we have much stricter guarantees on sample_element (gen might return something different every time - we also don't check for type equivalence). Are there any cases which depend on this (unintentional?) behavior? Though those should probably use MixedVec instead.

@chick it might also get useful to get rid of Data.allElements, which is already noted in the code comments as being something that is odd, because it also returns sample_elements. Perhaps as a follow-on PR.

File: src/test/scala/chiselTests/Direction.scala [hunk 3]
```diff
@@ -61,6 +61,72 @@ class DirectionSpec extends ChiselPropSpec with Matchers {
...
+      })
+    })
+  }
+
+  property("Empty directioned Bundles should not cause direction errors") {
+    an [Exception] should be thrownBy {
+      elaborate(new Module {
+        val io = IO(new Bundle {
+          val foo = UInt(8.W)
+          val x = Input(new Bundle {})
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/Direction.scala]: This test is doing the opposite of what it says, but I think what it says is correct

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala [hunk 4]
```diff
@@ -54,6 +54,10 @@ object SpecifiedDirection {
 sealed abstract class ActualDirection
 
 object ActualDirection {
+  /** The object does not exist / is empty and hence has no direction
+    */
+  case object Empty extends ActualDirection
```
Comment (jackkoenig) [inline chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Is this an API we want to expose or is it kind of an implementation detail?

**Comment by jackkoenig** [comment]
@ducky64 I think this looks good, thanks for cleaning it up. I do have a concern about 1 test and about the public existence of `ActualDirection.Empty`. 

Your observations are interesting and worth exploring (in follow on PRs of course 🙂). The indexing of unbound Vecs is a bit weird, but the use case of getting the types or directions of its elements is a real one (eg. to declare the type of a wire you want to connect to dynamic indexing). If we are to consider removing we must make a public API to get the sample element. Agreed that this API is better since it handles zero-length Vecs.

We should also investigate cloning from sample_element instead of gen, the former might have performance improvements **iff** we memoize autoclonetype. That being said, maybe someone is somehow finding utility with the current implementation...

File: src/test/scala/chiselTests/Direction.scala (hunk 3 above)
Comment (ducky64) [reply src/test/scala/chiselTests/Direction.scala]: derp, fixed

File: chiselFrontend/src/main/scala/chisel3/core/Data.scala (hunk 4 above)
Comment (ducky64) [reply chiselFrontend/src/main/scala/chisel3/core/Data.scala]: Fair point, it does add unnecessary complexity - let me see if I can refactor this out.

**Comment by ducky64** [comment]
@jackkoenig so I don't think I can eliminate the Empty direction without changes to the test cases. The primary issue is that in:
```scala
val io = IO(new Bundle {
  val foo = Input(UInt(8.W))
  val x = new Bundle {}
})
```
the Bundle has Unspecified direction, which doesn't match with the UInt having Specified (Input) direction. What the test logic seems to be going for is that "empty Bundles without an explicit direction from above (eg enclosing `Input(...)`) can behave as either Specified or Unspecified direction", which I don't see a good way of solving without inspecting into the contents of children (ugly, breaks abstraction) or adding an Empty direction.

One potential alternative is for Input/Output to set the ActualDirection of an empty Bundle when it has no children. It would work for:
```scala
val io = IO(new Bundle {
  val foo = Input(UInt(8.W))
  val x = Input(new Bundle { })
})
```
but break on:
```scala
val io = IO(new Bundle {
  val foo = Input(UInt(8.W))
  val x = new Bundle {
    val a = if (false) Input(UInt(8.W))
  }
})
```
since the Input is internal and an empty Bundle without an externally-set direction reverts to Unspecified.

**Comment by edwardcwang** [comment]
I feel that the following is a valid input that should be supported, especially in generators:

```scala
val io = IO(new Bundle {
  val x = new Bundle {
    val a = if (false) Input(UInt(8.W))
  }
})
```

**Comment by jackkoenig** [comment]
I can't approve since I'm original author, but I think this is good to go.
